hathor/consensus/poa/poa_block_producer.py

```diff
@@ -0,0 +1,173 @@
+from hathor.reactor import ReactorProtocol
+
+if TYPE_CHECKING:
+    from hathor.consensus.poa import PoaSigner
+    from hathor.manager import HathorManager
+    from hathor.transaction import Block
+    from hathor.transaction.poa import PoaBlock
+
+logger = get_logger()
+
+# Number of seconds to wait for a sync to finish before trying to produce blocks
+_WAIT_SYNC_DELAY: int = 30
+
+# Number of seconds used in random delay calculation
+_RANDOM_DELAY_MULTIPLIER: int = 1
+
+
+class PoaBlockProducer:
+    """
+    This class is analogous to mining classes, but for Proof-of-Authority networks.
+    It waits for blocks to arrive, gets templates, and propagates new blocks accordingly.
+    """
+    __slots__ = (
+        '_log',
+        '_settings',
+        '_poa_settings',
+        '_reactor',
+        '_manager',
+        '_poa_signer',
+        '_signer_index',
+        '_started_producing',
+        '_start_producing_lc',
+        '_schedule_block_lc',
+        '_last_seen_best_block',
+    )
+
+    def __init__(
+        self,
+        *,
+        settings: HathorSettings,
+        reactor: ReactorProtocol,
+        manager: 'HathorManager',
+        poa_signer: PoaSigner,
+    ) -> None:
+        assert isinstance(settings.CONSENSUS_ALGORITHM, PoaSettings)
+        self._log = logger.new()
+        self._settings = settings
+        self._poa_settings = settings.CONSENSUS_ALGORITHM
+        self._reactor = reactor
+        self._manager = manager
+        self._poa_signer = poa_signer
+        self._signer_index = self._calculate_signer_index(self._poa_settings, self._poa_signer)
+        self._last_seen_best_block: Block | None = None
+
+        self._started_producing = False
+        self._start_producing_lc = LoopingCall(self._start_producing)
+        self._start_producing_lc.clock = self._reactor
+
+        self._schedule_block_lc = LoopingCall(self._schedule_block)
+        self._schedule_block_lc.clock = self._reactor
+
+    def start(self) -> None:
+        self._start_producing_lc.start(_WAIT_SYNC_DELAY)
+        self._schedule_block_lc.start(self._settings.AVG_TIME_BETWEEN_BLOCKS)
+
+    def stop(self) -> None:
+        if self._start_producing_lc.running:
+            self._start_producing_lc.stop()
+
+        if self._schedule_block_lc.running:
+            self._schedule_block_lc.stop()
+
+    @staticmethod
+    def _calculate_signer_index(settings: PoaSettings, poa_signer: PoaSigner) -> int:
+        """Return the signer index for the given private key."""
+        public_key = poa_signer.get_public_key()
+        public_key_bytes = get_public_key_bytes_compressed(public_key)
+        sorted_signers = sorted(settings.signers)
+        try:
+            return sorted_signers.index(public_key_bytes)
+        except ValueError:
+            raise ValueError(f'Public key "{public_key_bytes.hex()}" not in list of PoA signers')
+
+    def _start_producing(self) -> None:
```

Comment:
You can remove this method and the `_schedule_block_lc` and replace it by simply calling `self._manager.can_start_mining()` at the beginning of the `_schedule_block` method.

Reply:
Done in a separate PR to make it easy to update future tests: https://github.com/HathorNetwork/hathor-core/pull/1080

Reply:
Ok! Thanks!